server/src/main/java/org/apache/iotdb/db/query/dataset/NewEngineDataSetWithoutValueFilter.java

```diff
@@ -0,0 +1,307 @@
+                }
+                ReadWriteIOUtils.write(floatValue, valueBAOSList[seriesIndex]);
+                break;
+              case DOUBLE:
+                double doubleValue = cachedBatchDataArray[seriesIndex].getDouble();
+                if (encoder != null) {
+                  doubleValue = encoder.encodeDouble(doubleValue, minTime);
+                }
+                ReadWriteIOUtils.write(doubleValue, valueBAOSList[seriesIndex]);
+                break;
+              case BOOLEAN:
+                ReadWriteIOUtils.write(cachedBatchDataArray[seriesIndex].getBoolean(),
+                    valueBAOSList[seriesIndex]);
+                break;
+              case TEXT:
+                ReadWriteIOUtils
+                    .write(cachedBatchDataArray[seriesIndex].getBinary(),
+                        valueBAOSList[seriesIndex]);
+                break;
+              default:
+                throw new UnSupportedDataTypeException(
+                    String.format("Data type %s is not supported.", type));
+            }
+          }
+
+          // move next
+          cachedBatchDataArray[seriesIndex].next();
+
+          // get next batch if current batch is empty
+          if (!cachedBatchDataArray[seriesIndex].hasCurrent()) {
+            if (seriesReaderWithoutValueFilterList.get(seriesIndex).hasNextBatch()) {
+              cachedBatchDataArray[seriesIndex] = seriesReaderWithoutValueFilterList
+                  .get(seriesIndex).nextBatch();
+            }
+          }
+
+          // try to put the next timestamp into the heap
+          if (cachedBatchDataArray[seriesIndex].hasCurrent()) {
+            timeHeap.add(cachedBatchDataArray[seriesIndex].currentTime());
+          }
+        }
+      }
+
+      if (rowOffset == 0) {
+        rowCount++;
+        if (rowCount % 8 == 0) {
+          for (int seriesIndex = 0; seriesIndex < seriesNum; seriesIndex++) {
+            ReadWriteIOUtils
+                .write((byte) currentBitmapList[seriesIndex], bitmapBAOSList[seriesIndex]);
+            // we should clear the bitmap every 8 row record
+            currentBitmapList[seriesIndex] = 0;
+          }
+        }
+        if (rowLimit > 0) {
+          alreadyReturnedRowNum++;
+        }
+      } else {
+        rowOffset--;
+      }
+    }
+
+    /*
+     * feed the bitmap with remaining 0 in the right
+     * if current bitmap is 00011111 and remaining is 3, after feeding the bitmap is 11111000
+     */
+    if (rowCount > 0) {
+      int remaining = rowCount % 8;
+      if (remaining != 0) {
+        for (int seriesIndex = 0; seriesIndex < seriesNum; seriesIndex++) {
+          ReadWriteIOUtils.write((byte) (currentBitmapList[seriesIndex] << (8 - remaining)),
+              bitmapBAOSList[seriesIndex]);
+        }
+      }
+    }
+
+    // set time buffer
+    ByteBuffer timeBuffer = ByteBuffer.allocate(timeBAOS.size());
+    timeBuffer.put(timeBAOS.toByteArray());
```

Comment:
We should use timeBAOS.getBuf() instead of timeBAOS.toByteArray().

Reply:
yes, fixed